src/lib/FreeC/Backend/Coq/Converter/TypeDecl.hs

```diff
@@ -177,23 +196,131 @@ convertDataDecl (IR.DataDecl _ (IR.DeclIdent _ name) typeVarDecls conDecls) = do
+    let typeVarIdents = map (fromJust . Coq.unpackQualid) typeVarQualids
+    returnType' <- convertType' returnType
+    -- Generate unqualified and qualified notation sentences for the smart
+    -- constructor if possible.
+    return
+      ( generateSmartConDecl' unqualIdent qualid typeVarIdents constrArgIdents
+          returnType'
+      , case mbModName of
+          Just modName -> generateSmartConDecl' (modName ++ '.' : unqualIdent)
+            qualid typeVarIdents constrArgIdents returnType'
+          Nothing      -> []
+      )
+
+  -- | Generates a notation sentence for the smart constructor and the
+  --   explicit application of the smart constructor.
+  generateSmartConDecl'
+    :: String
+    -> Coq.Qualid
+    -> [String]
+    -> [String]
+    -> Coq.Term
+    -> [Coq.Sentence]
+  generateSmartConDecl' smartIdent constr typeVarIdents constrArgIdents
+    expReturnType = [ Coq.notationSentence lhs rhs mods
+                    , Coq.notationSentence expLhs expRhs expMods
+                    ]
+   where
+    freeArgIdents = map (fromJust . Coq.unpackQualid . fst) Coq.Base.freeArgs
+
+    freeArgs      = map (Coq.Qualid . fst) Coq.Base.freeArgs
+
+    -- Default smart constructor with implicit type args.
+    returnType    = case expReturnType of
+      (Coq.App tcon (shape NonEmpty.:| pos : tvars))
+        | length tvars == length typeVarIdents -> Coq.App tcon
+          (shape NonEmpty.:| pos
+           : map (Coq.PosArg . const Coq.Underscore) tvars)
+      _ -> Coq.Underscore
+
+    argIdents     = freeArgIdents ++ constrArgIdents
+
+    args          = freeArgs
+      ++ map (const Coq.Underscore) typeVarIdents
+      ++ map (Coq.Qualid . Coq.bare) constrArgIdents
+
+    lhs           = Coq.nSymbol smartIdent NonEmpty.:| map Coq.nIdent argIdents
+
+    rhs           = Coq.explicitApp Coq.Base.freePureCon
+      $ freeArgs ++ [returnType, Coq.explicitApp constr args]
+
+    mods          = [ Coq.sModLevel 10
+                    , Coq.sModIdentLevel (NonEmpty.fromList argIdents) (Just 9)
+                    ]
+
+    -- Explicit notation for the smart constructor.
+    expArgIdents  = freeArgIdents ++ typeVarIdents ++ constrArgIdents
+
+    expArgs       = map (Coq.Qualid . Coq.bare) expArgIdents
+
+    expLhs        = Coq.nSymbol ('@' : smartIdent)
+      NonEmpty.:| map Coq.nIdent expArgIdents
+
+    expRhs        = Coq.explicitApp Coq.Base.freePureCon
+      $ freeArgs ++ [expReturnType, Coq.explicitApp constr expArgs]
+
+    expMods
+      = [ Coq.SModOnlyParsing
+        , Coq.sModLevel 10
+        , Coq.sModIdentLevel (NonEmpty.fromList expArgIdents) (Just 9)
+        ]
 -- Type synonyms are not allowed in this function.
 convertDataDecl (IR.TypeSynDecl _ _ _ _)
   = error "convertDataDecl: Type synonym not allowed."
+
+-- | Partitions a list of sentences into a list of sentences that belong to
+--   qualified smart constructors and a list of the remaining sentences.
+partitionIsQualifiedSmartConstructor
```

Comment:
This seems like a somewhat roundabout way of separating qualified notations from other sentences. 
If we ever decide to add more (unrelated) notations to the translation of a type declaration, and happen to put them after the qualified notations, they would also be counted as qualified notations.

Since `generateSmartConDecls` already returns the qualified notations separately, couldn't they just get passed through to `convertTypeComponent`, i.e. couldn't `convertTypeComponent` return a pair of `Sentence` lists, with the first list containing the data type declarations, unqualified notations etc. and the second list containing the qualified notations?

Reply:
I wasn't sure which of those implementations was the best way. I decided to use the partition function to have few changes to the signatures of existing functions and keep the interface to the type conversion simple. But I guess your right that the explicit separation of those sentences through the code is more fail-safe regarding future changes and should therefore be used.